pallets/delegation/src/migrations.rs

```diff
@@ -0,0 +1,480 @@
+	#[allow(clippy::absurd_extreme_comparisons)]
+	pub(crate) fn pre_migration(&self) -> Result<(), DelegationMigrationError> {
+		ensure!(
+			LastUpgradeVersion::<T>::get() <= migrations::LATEST_UPGRADEABLE_VERSION,
+			DelegationMigrationError::AlreadyLatest
+		);
+
+		Ok(())
+	}
+
+	// It retrieves the latest version deployed on chain, and sequencially applies
+	// any runtime migration until the state is update to the latest version.
+	// It provides a test feature that panics whenver something goes wrong for one
+	// of the version migrators, but in production it is just assumed that all the
+	// migrations will go through.
+	pub(crate) fn migrate(&self) -> Weight {
+		let mut total_weight_used: Weight = 0;
+		let current_version = LastUpgradeVersion::<T>::get();
+		// Applies all the needed migration from the current version up to
+		// LATEST_UPGRADEABLE_VERSION (which is always the latest possible version - 1).
+		for version in current_version..=LATEST_UPGRADEABLE_VERSION {
+			// It is assumed that the are exactly LATEST_UPGRADEABLE_VERSION+1 migrators, so
+			// as long as the pre-condition is met, the following is assumed to never panic.
+			let version_migrator: &dyn VersionMigrator<T> = self.migrations[version as usize].as_ref();
+			// Test pre-conditions (only in testing) for each version migrator.
+			#[cfg(feature = "try-runtime")]
+			if let Err(err) = version_migrator.pre_migrate() {
+				panic!("{:?}", err);
+			}
+			// Apply the migration and keep adding the weight.
+			total_weight_used = total_weight_used.saturating_add(version_migrator.migrate());
+			// Test post-conditions (only in testing) for each version migrator.
+			#[cfg(feature = "try-runtime")]
+			if let Err(err) = version_migrator.post_migrate() {
+				panic!("{:?}", err);
+			}
+		}
+		// Set a version number that is not upgradeable anymore until a new version is
+		// available, which means that the new version will be
+		// LATEST_UPGRADEABLE_VERSION + 1.
+		LastUpgradeVersion::<T>::set(LATEST_UPGRADEABLE_VERSION.saturating_add(1));
+
+		// Add a DB read and write for the LastUpgradeVersion storage update
+		total_weight_used.saturating_add(T::DbWeight::get().reads_writes(1, 1))
+	}
+
+	// Checks whether the upgrade as a whole went through by verifying that the
+	// latest version set it not upgradeable anymore.
+	#[cfg(any(feature = "try-runtime", test))]
+	pub(crate) fn post_migration(&self) -> Result<(), DelegationMigrationError> {
+		ensure!(
+			LastUpgradeVersion::<T>::get() == migrations::LATEST_UPGRADEABLE_VERSION.saturating_add(1),
+			DelegationMigrationError::MigrationResultInconsistent
+		);
+
+		Ok(())
+	}
+}
+
+mod v0 {
+	use super::*;
+	// Migrator for the first actual pallet's migration. It migrates the Roots,
+	// Delegations, and Children storage entries to a simpler HierarchyInfos,
+	// DelegationNodes, while maintaining the original information an all
+	// parent-child links.
+	pub struct V0Migrator();
```

Comment:
Can't we just call this `Migrator`? It's already versioned by the module.
```suggestion
	pub struct Migrator();
```

Reply:
Changed almost everything in the new migrations module.